crates/sparrow-instructions/src/evaluators/list/index.rs

```diff
@@ -0,0 +1,189 @@
+use sparrow_plan::ValueRef;
+use std::sync::Arc;
+
+use crate::{Evaluator, EvaluatorFactory, StaticInfo};
+
+/// Evaluator for `index` on lists.
+///
+/// Retrieves the value at the given index.
+#[derive(Debug)]
+pub(in crate::evaluators) struct IndexEvaluator {
+    index: ValueRef,
+    list: ValueRef,
+}
+
+impl EvaluatorFactory for IndexEvaluator {
+    fn try_new(info: StaticInfo<'_>) -> anyhow::Result<Box<dyn Evaluator>> {
+        let input_type = info.args[1].data_type.clone();
+        match input_type {
+            DataType::List(t) => anyhow::ensure!(t.data_type() == info.result_type),
+            other => anyhow::bail!("expected list type, saw {:?}", other),
+        };
+
+        let (index, list) = info.unpack_arguments()?;
+        Ok(Box::new(Self { index, list }))
+    }
+}
+
+impl Evaluator for IndexEvaluator {
+    fn evaluate(&mut self, info: &dyn crate::RuntimeInfo) -> anyhow::Result<ArrayRef> {
+        let list_input = info.value(&self.list)?.list_array()?;
+        let index_input = info.value(&self.index)?.primitive_array()?;
+
+        let result = list_get(&list_input, &index_input)?;
+        Ok(Arc::new(result))
+    }
+}
+
+/// Given a `ListArray` and `index` array of the same length return an array of the values.
+fn list_get(list: &ListArray, indices: &Int64Array) -> anyhow::Result<ArrayRef> {
+    anyhow::ensure!(list.len() == indices.len());
+    let take_indices = list_indices(list, indices)?;
+    arrow::compute::take(list.values(), &take_indices, None).context("take in get_map")
+}
+
+/// Gets the indices in the list where the values are at the index within each list.
+fn list_indices(list: &ListArray, indices: &Int64Array) -> anyhow::Result<Int32Array> {
+    let offsets = list.offsets();
+    let values = list.values();
+    downcast_primitive_array!(
+        values => {
+               Ok(accessible_array_list_indices(
+                offsets,
+                values,
+                indices
+            ))
+        }
+        DataType::Utf8 => {
+            let values = as_string_array(values);
+            Ok(accessible_array_list_indices(
+                offsets,
+                values,
+                indices
+            ))
+        }
+        DataType::LargeUtf8 => {
+            let values = as_largestring_array(values);
+            Ok(accessible_array_list_indices(
+                offsets,
+                values,
+                indices
+            ))
+        }
+        DataType::Boolean => {
+            let values = as_boolean_array(values);
+            Ok(accessible_array_list_indices(
+                offsets,
+                values,
+                indices
+            ))
+        }
+        unsupported => {
+            anyhow::bail!("unsupported list type: {:?}", unsupported)
+        }
+    )
+}
+
+/// Generic implementation of `list_indices` for arrays implementing `ArrayAccessor`.
+fn accessible_array_list_indices<T: Send + Sync, A: ArrayAccessor<Item = T>>(
```

Comment:
This doesn't need the values, which means it doesn't need to know the type. Specifically, if `value[computed_index]` is null, then `take(values, computed_index)` will be null. So we can *just* compute the index.

With map we needed to handle the case where the *key* was null, but that doesn't happen here. I wouldn't expect list indexing to need to care about the type (or nullness) of values in the list, since it just works with the indices.

Reply:
Ahh okay, I thought so but for some reason thought `take` would get the `value` regardless of nullness